tooling/noirc_abi/src/arbitrary.rs

```diff
@@ -0,0 +1,154 @@
+            value.push('1');
+        }
+        seen_values.insert(value.clone());
+    }
+}
+
+proptest::prop_compose! {
+    pub(super) fn arb_field_from_integer(bit_size: u32)(value: u128)-> FieldElement {
+        let width = (bit_size % 128).clamp(1, 127);
+        let max_value = 2u128.pow(width) - 1;
+        FieldElement::from(value.clamp(0, max_value))
+    }
+}
+
+fn arb_value_from_abi_type(abi_type: &AbiType) -> SBoxedStrategy<InputValue> {
+    match abi_type {
+        AbiType::Field => vec(any::<u8>(), 32)
+            .prop_map(|bytes| InputValue::Field(FieldElement::from_be_bytes_reduce(&bytes)))
+            .sboxed(),
+        AbiType::Integer { width, .. } => {
+            arb_field_from_integer(*width).prop_map(InputValue::Field).sboxed()
+        }
+
+        AbiType::Boolean => {
+            any::<bool>().prop_map(|val| InputValue::Field(FieldElement::from(val))).sboxed()
+        }
+
+        AbiType::String { length } => {
+            // Strings only allow ASCII characters as each character must be able to be represented by a single byte.
+            let string_regex = format!("[[:ascii:]]{{{length}}}");
+            proptest::string::string_regex(&string_regex)
+                .expect("fix me")
+                .prop_map(InputValue::String)
+                .sboxed()
+        }
+        AbiType::Array { length, typ } => {
+            let length = *length as usize;
+            let elements = vec(arb_value_from_abi_type(typ), length..=length);
+
+            elements.prop_map(InputValue::Vec).sboxed()
+        }
+
+        AbiType::Struct { fields, .. } => {
+            let fields: Vec<SBoxedStrategy<(String, InputValue)>> = fields
+                .iter()
+                .map(|(name, typ)| (Just(name.clone()), arb_value_from_abi_type(typ)).sboxed())
+                .collect();
+
+            fields
+                .prop_map(|fields| {
+                    let fields: BTreeMap<_, _> = fields.into_iter().collect();
+                    InputValue::Struct(fields)
+                })
+                .sboxed()
+        }
+
+        AbiType::Tuple { fields } => {
+            let fields: Vec<_> = fields.iter().map(arb_value_from_abi_type).collect();
+            fields.prop_map(InputValue::Vec).sboxed()
+        }
+    }
+}
+
+fn arb_primitive_abi_type() -> SBoxedStrategy<AbiType> {
+    const MAX_STRING_LEN: u32 = 1000;
+    proptest::prop_oneof![
+        Just(AbiType::Field),
+        Just(AbiType::Boolean),
+        any::<(Sign, u32)>().prop_map(|(sign, width)| {
+            let width = (width % 128).clamp(1, 127);
+            AbiType::Integer { sign, width }
+        }),
+        // restrict length of strings to avoid running out of memory
+        (1..MAX_STRING_LEN).prop_map(|length| AbiType::String { length }),
+    ]
+    .sboxed()
+}
+
+pub(super) fn arb_abi_type() -> BoxedStrategy<AbiType> {
+    let leaf = arb_primitive_abi_type();
+
+    leaf.prop_recursive(
+        8,   // 8 levels deep
```

Comment:
nit: `up to 8 levels deep`, had to check docs to see whether it'd ever be `0` levels deep (i.e. a primitive type)